README.md
# Docker Toran Proxy

[![Version](https://images.microbadger.com/badges/version/cedvan/toran-proxy:1.5.4-2.svg)](https://microbadger.com/images/cedvan/toran-proxy:1.5.4-2 "Get your own version badge on microbadger.com")
[![Docker Pulls](https://img.shields.io/docker/pulls/cedvan/toran-proxy.svg?style=flat-square)](https://hub.docker.com/r/cedvan/toran-proxy/)
[![Docker Stars](https://img.shields.io/docker/stars/cedvan/toran-proxy.svg?style=flat-square)](https://hub.docker.com/r/cedvan/toran-proxy/)
[![Docker image](https://images.microbadger.com/badges/image/cedvan/toran-proxy.svg?style=flat-square)](https://microbadger.com/#/images/cedvan/toran-proxy "Size docker image on Docker Hub")
[![Build Status](https://img.shields.io/travis/cedvan/docker-toran-proxy/master.svg?style=flat-square)](https://travis-ci.org/cedvan/docker-toran-proxy)
[![Join the chat at https://gitter.im/cedvan/docker-toran-proxy](https://badges.gitter.im/cedvan/docker-toran-proxy.svg)](https://gitter.im/cedvan/docker-toran-proxy?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![GitHub license](https://img.shields.io/:license-mit-blue.svg?style=flat-square)]()

![toran-proxy](https://raw.githubusercontent.com/cedvan/docker-toran-proxy/master/img/toran-proxy.png "Toran-Proxy")

---

:warning: | **Toran proxy stating no longer supported and that people should use Private Packagist at https://packagist.com instead**
------------ | -------------

---

Toran acts as a proxy for Packagist and GitHub. It is meant to be set up on your own server or even inside your office. This offers a few benefits:

- **Redundant infrastructure to ensure your deployments never fail and your developers can work at any time.** Packages will be installed from your proxy with a fallback to GitHub, ensuring a maximum availability.
- **Higher bandwidth for faster installations.** You can set up Toran in your local network or on a server near you.

## Quick start

```bash
docker run --name toran-proxy -d \
    -p 80:80 \
    cedvan/toran-proxy:1.5.4-2
```
Go with your browser to **localhost**

## Save data

Files are saved to `/data/toran-proxy` in container. Just mount this volume for save your configurations and repositories

```bash
docker run --name toran-proxy -d \
    -v /opt/toran-proxy:/data/toran-proxy \
    cedvan/toran-proxy:1.5.4-2
```

## Add ssh config for private repository

```bash
docker run --name toran-proxy -d \
    -p 443:443 \
    -v /opt/toran-proxy/ssh:/data/toran-proxy/ssh \
    cedvan/toran-proxy:1.5.4-2
```
*Files supported : `id_rsa`, `id_rsa.pub`, `config` and `known_hosts`*

## Configure Cron timer

```bash
docker run --name toran-proxy -d \
    -p 443:443 \
    -e "TORAN_CRON_TIMER=half" \
    cedvan/toran-proxy:1.5.4-2
```

## Enabled HTTPS

```bash
docker run --name toran-proxy -d \
    -p 443:443 \
    -e "TORAN_HTTPS=true" \
    -v /opt/toran-proxy/certs:/data/toran-proxy/certs \
    cedvan/toran-proxy:1.5.4-2
```
Add your **toran-proxy.key** and **toran-proxy.crt** in folder **certs**. If `toran-proxy.key` and `toran-proxy.crt` do not exist, the container will create self-signed certificates

## HTTP Authentification

Use file **htpasswd** to add authentification ?

Add `auth.json` to composer configuration home folder

```
{
    "http-basic": {
        "toran-proxy.domain.tld": {
            "username": "myUsername",
            "password": "myPassword"
        },
    }
}
```

## Toran Proxy Options

*Please refer the docker run command options for the `--env-file` flag where you can specify all required environment variables in a single file. This will save you from writing a potentially long docker run command. Alternately you can use fig.*

Below is the complete list of available options that can be used to customize your toran proxy installation.

- **TORAN_HOST**: The hostname of the toran proxy server. Defaults to `localhost`
- **TORAN_HTTP_PORT**: The port of the toran http server. Defaults to `80`
- **TORAN_HTTPS**: Set to `true` to enable https support, Defaults to `false`. **If you do not use a reverse proxy, do not forget to add the certificates files**
- **TORAN_HTTPS_PORT**: The port of the toran https server. Defaults to `443`
- **TORAN_REVERSE**: Set to `true` if you use docker behind a reverse proxy for i.e. ssl termination. This will make Toran use the HTTPS scheme without the need to add certificates. If you do so, make sure to set your reverse proxy to target port 443. Defaults to `false`
- **TORAN_CRON_TIMER**: Setup cron job timer. Defaults to `fifteen`
    - `minutes`: All minutes
    - `five`: All five minutes
    - `fifteen`: All fifteen minutes
    - `half`: All thirty minutes
    - `hour`: All hours
    - `daily`: All days at 04:00 (Use *TORAN_CRON_TIMER_DAILY_TIME* for customize time)
- **TORAN_CRON_TIMER_DAILY_TIME**: Set a time for cron job daily timer in `HH:MM` format. Defaults to `04:00`
- **TORAN_TOKEN_GITHUB**: Add your Github token for ensure download repositories since Github. Default `null`.
- **TORAN_TRACK_DOWNLOADS**: Track private package installs, set to `true` to get an install log in `/data/toran-proxy/logs/downloads.private.log`. Defaults to `false`
- **TORAN_MONO_REPO**: Set to `true` for switch to a monorepo instead of dual repo model. Defaults to `false`
- **PHP_TIMEZONE**: Configure timezone PHP. Default `Europe/Paris`.
- **TORAN_AUTH_ENABLE**: Set to `true` to enable HTTP Basic Authentication. When enabled, `TORAN_AUTH_USER` and `TORAN_AUTH_PASSWORD` are required. Defaults to `false`.
- **TORAN_AUTH_USER**: Configure the HTTP Basic Authentication Username. Defaults to `toran`.
- **TORAN_AUTH_PASSWORD**: Configure the HTTP Basic Authentication Password. Defaults to `toran`.

## Add customization scripts

For scenarios where the degree of configurability his image offers via the
above listed options is not sufficient, you are able to add container local
customization scripts which will get executed during container runtime. Here
you can add for example sed calls which further tweak the nginx configuration.

The container `launch.sh` script expects custom scripts to be found under
`/data/toran-proxy/scripts/*.sh`. These scripts just get sourced in order.

```bash
export custdir=/tmp/toran-customs
mkdir -p $custdir
echo "echo 'hello world'" > $custdir/hello.sh
docker run --name toran-proxy -d \
    -p 443:443 \
    -v $custdir:/data/toran-proxy/scripts
    cedvan/toran-proxy:1.5.4-2
```

## Toran Proxy License

By default, Toran proxy license is for personal use.
You can add a license from the Toran proxy interface

## Contributing

See [CONTRIBUTING](CONTRIBUTING.md) file.

## References

Toran is built by Jordi Boggiano, lead developer of Composer. As such he can make sure they work well together. No surprises.

- https://toranproxy.com/
- https://twitter.com/toranproxy
- https://getcomposer.org/doc/articles/handling-private-packages-with-satis.md
- https://github.com/jwilder/nginx-proxy


Changelog.md
# Changelog

**latest**
- Enable HTTP/2 in Nginx.

**1.5.4-2**
- Upgrade Ubuntu 18.04 to 20.04
- Upgrade PHP 7.2 to 7.4

**1.5.4-1**
- Derive `TORAN_SCHEME` from `TORAN_HTTPS` resp. `TORAN_HTTP` which allows toran to spit out HTTPS links despite being addressed via HTTP (reverse proxy scenario)
- Upgrade PHP 7.1 to PHP 7.2
- Remove unnecessary PPA
- Upgrade Ubuntu 14.04 to 18.04
- Optimize layers by doing all apt and purge as a single run. (Down from 499MB to 355MB)
- Fix the tarball extract and copy

**1.5.4**
- Add custom scripts directory `/data/toran-proxy/scripts` for user customizations to be executed on startup
- Update FROM to `ubuntu:14.04`
- Upgrade to PHP 7.1
- Upgrade toran proxy to 1.5.4

**1.5.3-1**
- Enabling HTTP Basic Authentication configuration with `TORAN_AUTH_ENABLE`, `TORAN_AUTH_USER`, and `TORAN_AUTH_PASSWORD` env vars.
- Fix load toran logs
- Fix permissions on packagist directory

**1.5.3**
- Upgrade toran proxy to version 1.5.3
- Fixing wrong declaration of cron "hour" value.

**1.5.2**
- Upgrade toran proxy to version 1.5.2
- Redirection `http` to `https` auto with use https vhost type
- Generate self-signed certificates automatically
- Clean create and permissions on logs directory
- Add option `TORAN_HTTP_PORT` to allow arbitrary http port for nginx / toran proxy
- Add option `TORAN_HTTPS_PORT` to allow arbitrary https port for nginx / toran proxy
- Optimize check environment configuration

**1.5.1**
- Upgrade toran proxy to version 1.5.1
- Fix update github token

**1.5.0**
- Upgrade toran proxy to version 1.5.0
- Add `monorepo` to `config.yml` of toran proxy
- Add option `TORAN_TRACK_DOWNLOADS` to track private package installs with logs
- Add option `TORAN_MONO_REPO` to track private package installs with logs

**1.4.4**
- Upgrade toran proxy to version 1.4.4

**1.4.3**
- Upgrade toran proxy to version 1.4.3

**1.4.2**
- Upgrade toran proxy to version 1.4.2

**1.4.1**
- Upgrade toran proxy to version 1.4.1

**1.4.0**
- Upgrade toran proxy to version 1.4.0
- Add toran proxy logs to data volume saved
- Add `track_downloads` to `config.yml` of toran proxy
- Add a contributing file

**1.3.2**
- Upgrade toran proxy to version 1.3.2
- Fix create packages directory
- Fix bad restart synchronisation
- Add microbadger to report docker image size
- Add option `TORAN_REVERSE` to running toran proxy behind a reverse proxy for i.e. SSL termination.

**1.2.0**
- Optimize README.md
- Upgrade toran proxy to version 1.2.0
- Global clean
- Add support `ssh/config` for private repository
- Rename environment variable `TORAN_PHP_TIMEZONE` to `PHP_TIMEZONE`
- Change user nginx and php-fpm from `root` to `www-data`
- Add `ssh` package
- Upgrade image base to `cedvan/ubuntu:14.04.20160326`

**1.1.7-2**
- Add supervisor
- Redirect logs to stdout
- Start first sync on startup
- Change `TORAN_CRON_TIMER` to `fifteen` by default
- Add PHP logs

**1.1.7-1**
- Fix sed in toran.sh
- Add Cron timer configuration
- Optimize logs directories
- Remove launch cron on start to speed up launch
- Optimize loading permissions
- Fix bug with token github empty

**1.1.7**
- Update Toran Proxy to version 1.1.7

**1.1.6-3**
- Update image FROM to cedvan/ubuntu:14.04.20150414
- Add ssh config for root user
- Fix License
- Fix config github token for composer

**1.1.6-2**
- Add **cron** every minute for download packages in background
- Fix toran config `git_path`
- Optimize logs directories

**1.1.6-1**
- Download Toran Proxy in Dockerfile
- Move save data in `/data/toran-proxy`
- Add default config for *git* and save mirrors in /data
- Rename `settings.yml` to `config.yml`
- Remove useless `parameters.yml`
- Move `assets/config` to `/assets/config` in container
- Remove useless `init.sh`
- Remove useless configuration *TORAN_PACKAGIST*, may change with toran interface
- Remove useless configuration *TORAN_SYNC*, may change with toran interface

**1.1.6**
- First version


scripts/config.sh
#!/bin/bash

echo "Check environment configuration..."

# Directories
DATA_DIRECTORY=/data/toran-proxy
WORK_DIRECTORY=/var/www
ASSETS_DIRECTORY=/assets
SCRIPTS_DIRECTORY=/scripts/toran-proxy

# PHP Configuration
PHP_TIMEZONE=${PHP_TIMEZONE:-Europe/Paris}

# Toran Proxy Configuration
TORAN_HOST=${TORAN_HOST:-localhost}
TORAN_HTTP_PORT=${TORAN_HTTP_PORT:-80}
TORAN_HTTPS=${TORAN_HTTPS:-false}
TORAN_HTTPS_PORT=${TORAN_HTTPS_PORT:-443}
TORAN_REVERSE=${TORAN_REVERSE:-false}
TORAN_CRON_TIMER=${TORAN_CRON_TIMER:-fifteen}
TORAN_CRON_TIMER_DAILY_TIME=${TORAN_CRON_TIMER_DAILY_TIME:-04:00}
TORAN_TOKEN_GITHUB=${TORAN_TOKEN_GITHUB:-false}
TORAN_TRACK_DOWNLOADS=${TORAN_TRACK_DOWNLOADS:-false}
TORAN_MONO_REPO=${TORAN_MONO_REPO:-false}
TORAN_SECRET=$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1)
TORAN_AUTH_ENABLE=${TORAN_AUTH_ENABLE:-false}
TORAN_AUTH_USER=${TORAN_AUTH_USER:-toran}
TORAN_AUTH_PASSWORD=${TORAN_AUTH_PASSWORD:-toran}

# Checking PHP Timezone
if [[ ! "${PHP_TIMEZONE}" =~ ^[A-Z]{1}[a-z]+/[A-Z]{1}[a-z]+$ ]]; then
    echo "ERROR: "
    echo "  Variable PHP_TIMEZONE isn't valid ! (Format accepted : [A-Z]{1}[a-z]+/[A-Z]{1}[a-z]+)"
    exit 1
fi

# Checking Toran HTTPS Configuration and load Toran scheme
if [ "${TORAN_HTTPS}" = "true" ]; then
    [ -z "${TORAN_SCHEME}" ] && TORAN_SCHEME="https"
elif [ "${TORAN_HTTPS}" = "false" ]; then
    [ -z "${TORAN_SCHEME}" ] && TORAN_SCHEME="http"
else
    echo "ERROR: "
    echo "  Variable TORAN_HTTPS isn't valid ! (Values accepted : true/false)"
    exit 1
fi

# Checking HTTP PORT
if [[ ! ("${TORAN_HTTP_PORT}" =~ ^[1-9]{1}[0-9]{0,4}$) ]] || (("${TORAN_HTTP_PORT}" > "65535")); then
    echo "ERROR: "
    echo "  Variable TORAN_HTTP_PORT isn't valid ! (Format accepted : integer between 1 and 65535)"
    exit 1
fi

# Checking HTTPS PORT
if [[ ! ("${TORAN_HTTPS_PORT}" =~ ^[1-9]{1}[0-9]{0,4}$) ]] || (("${TORAN_HTTPS_PORT}" > "65535")); then
    echo "ERROR: "
    echo "  Variable TORAN_HTTP_PORT isn't valid ! (Format accepted : integer between 1 and 65535)"
    exit 1
fi

# Checking Toran reverse
if [ "${TORAN_REVERSE}" != "true" ] && [ "${TORAN_REVERSE}" != "false" ]; then
    echo "ERROR: "
    echo "  Variable TORAN_REVERSE isn't valid ! (Values accepted : true/false)"
    exit 1
fi

# Checking Toran track reverse
if [ "${TORAN_TRACK_DOWNLOADS}" != "true" ] && [ "${TORAN_TRACK_DOWNLOADS}" != "false" ]; then
    echo "ERROR: "
    echo "  Variable TORAN_TRACK_DOWNLOADS isn't valid ! (Values accepted : true/false)"
    exit 1
fi

# Checking Toran mono repo
if [ "${TORAN_MONO_REPO}" != "true" ] && [ "${TORAN_MONO_REPO}" != "false" ]; then
    echo "ERROR: "
    echo "  Variable TORAN_MONO_REPO isn't valid ! (Values accepted : true/false)"
    exit 1
fi

# Checking Toran cron timer
if [ "${TORAN_CRON_TIMER}" != "minutes" ] && [ "${TORAN_CRON_TIMER}" != "five" ] && [ "${TORAN_CRON_TIMER}" != "fifteen" ] && [ "${TORAN_CRON_TIMER}" != "half" ] && [ "${TORAN_CRON_TIMER}" != "hour" ] && [ "${TORAN_CRON_TIMER}" != "daily" ]; then
    echo "ERROR: "
    echo "  Variable TORAN_CRON_TIMER isn't valid ! (Values accepted : minutes/five/fifteen/half/hour)"
    exit 1
fi

# Checking Toran cron daily time
if [[ ! "${TORAN_CRON_TIMER_DAILY_TIME}" =~ ^[0-9]{2}:[0-9]{2}$ ]]; then
    echo "ERROR: "
    echo "  Variable TORAN_CRON_TIMER_DAILY_TIME isn't valid ! (Format accepted : HH:MM)"
    exit 1
fi

# Checking Toran Auth Enable
if [ "${TORAN_AUTH_ENABLE}" != "true" ] && [ "${TORAN_AUTH_ENABLE}" != "false" ]; then
    echo "ERROR: "
    echo "  Variable TORAN_AUTH_ENABLE isn't valid ! (Values accepted : true/false)"
    exit 1
fi

# Checking Toran Auth User
if [ "${TORAN_AUTH_USER}" == "" ]; then
    echo "ERROR: "
    echo "  Variable TORAN_AUTH_USER isn't valid! Must not be empty."
    exit 1
fi

# Checking Toran Auth Password
if [ "${TORAN_AUTH_PASSWORD}" == "" ]; then
    echo "ERROR: "
    echo "  Variable TORAN_AUTH_PASSWORD isn't valid! Must not be empty."
    exit 1
fi
